### ISSUE TITLE
BUG: VR-Direct mode on AMD GPU cannot set refresh rate

### ISSUE THREAD

**Issue by TheDeveloperGuy** [ValveSoftware/openvr#1276]
Normally when a HMD driver wishes to specify the refresh rate for the display it's done via the Prop_DisplayFrequency_Float property. This is ignored for VR-Direct mode, for some non-consistent reason, and so the driver needs to set the "preferredRefreshRate" key in the "steamvr" section of the steamvr.vrsettings file to get SteamVR to create the correct VR-Direct screen mode. 

This clunky mechanism works for nVidia GPUs, but when attempting to do this for AMD GPUs, SteamVR simply ignores the attempt to set the key (you can break directly after the settings "sync" and see the file is not updated, even if a "sync" is forced), and also ignores any attempt to set the key manually (ie. by editing the config file in a text editor). This means for a headset that supports multiple refresh rates the highest refresh rate is taken and any attempt to specify it is ignored. 

Obviously this is not what anyone attempting to set this key, either programatically or manually is intending, so why do the correct thing for nVidia and not AMD?

**Comment by TheDeveloperGuy** [comment]
It seems the cause of this issue is in the EDID querying of the headset. If I go into the adapter settings for the HMD display in Windows, all refresh rates are listed. All refresh rates are also available in SteamVR when running in 'extended' mode. All refresh rates are also available in SteamVR when running the HMD on an nVidia GPU, in either 'extended' or 'direct' mode. Wherever the EDID is being queried for 'direct' mode, something is going wrong, as it's only picking up the highest refresh rate and ignoring the others. This has the effect of the settings key being ignored and a single refresh rate being visible in SteamVR general settings. Not sure what's going wrong there as Index HMD does not have this problem (on AMD).

**Comment by TheDeveloperGuy** [comment]
Seems that SteamVR will remove the settings key (or skip writing it) for the 'native' refresh rate of the HMD, on either nVidia or AMD, so the issue is something is going wrong in the EDID query for AMD. It appears everything else would work if that could be fixed.

**Comment by TheDeveloperGuy** [comment]
In the meantime, it seems I can fix this issue by changing the EDID returned so that it only specifies the refresh rate the user has requested. This also has the side-effect of removing the refresh rate selection from the SteamVR general settings, which simplifies things, as I then don't need to track changes made to the "preferred" refresh rate there.

**Comment by TheDeveloperGuy** [comment]
This doesn't appear to be an issue on Index on AMD, which also supports multiple refresh rates. In any case, I've worked around this by changing the EDID returned by the headset. It's worth looking at in case there is an actual bug in the OpenVR code. However, as VR-Direct on AMD has all kinds of weirdness that I've had to work around, I don't know how much of this is the fault of the AMD driver itself. Feel free to close.
